### PR TITLE
adding sfcship test

### PR DESCRIPTION
**Pull request by AndrewEichmann-NOAA**
Adding sfcship test

Files changed:
test/CMakeLists.txt
test/soca/CMakeLists.txt

Files added:
test/testinput/bufr_sfcshp.yaml

### CODE REVIEW

File: test/testinput/bufr_sfcshp.yaml [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# (C) Copyright 2020 NOAA/NWS/NCEP/EMC
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+
```
Comment (guillaumevernieres) [inline test/testinput/bufr_sfcshp.yaml]: Remove License header

File: test/testinput/bufr_sfcshp.yaml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+observations:
+  - obs space:
+      name: bufr
+      #obsdatain: "./testdata/bufr_sfcshp.bufr"
+      #obsdatain: "./testdata/shipsu.20180415_subsampled"
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshp.yaml]: Remove commented lines

**Review by ShastriPaturi — CHANGES_REQUESTED** [review]

File: test/testinput/bufr_sfcshp.yaml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+          t-profile:
+            query: "*/IDMSMDBS/BBYSTSL/SST1"
+
+          s-profile:
+            query: "*/IDMSMDBS/BBYSTSL/SALN"
+
+          depth:
+            query: "*/IDMSMDBS/BBYSTSL/DBSS"
+
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshp.yaml]: Remove t-profile, s-profile and depth. You are looking at surface data.

**Review by ShastriPaturi — CHANGES_REQUESTED** [review]

File: test/testinput/bufr_sfcshp.yaml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+          longitude:
+            query: "*/CLONH"
+
+          latitude:
+            query: "*/CLATH"
+
+          sst:
+            query: "*/SST0"
+
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshp.yaml]: to convert from K to degC add:
            transforms:
              - offset: -273.15
  
below: query: "*/SST0"

Comment (ShastriPaturi) [reply test/testinput/bufr_sfcshp.yaml]: Please query MSST (This is the method of SST measurement and is important for adding ObsError values. On ship, SST is measured using 2/3 locations, e.g ERI (Engine Roome Intake), Hull, bucket)

File: test/testinput/bufr_sfcshp.yaml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+          source: variables/depth
+          longName: "Water Depth Below Surface"
+          units: "m"
+
+        - name: "waterTemperature@ObsValue"
+          coordinates: "longitude latitude Levels"
+          source: variables/t-profile
+          longName: "Sea Water Temperature"
+          units: "K"
+          range: [250, 320]
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshp.yaml]: Lines 90-101 are not necessary and the variable names are incorrect.

File: test/testinput/bufr_sfcshp.yaml [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+          units: "degrees_east"
+          range: [-180, 180]
+
+        - name: "seaSurfaceTemperature@ObsValue"
+          coordinates: "longitude latitude"
+          source: variables/sst
+          longName: "Sea Surface Temperature"
+          units: "K"
+          range: [250, 320]
+
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshp.yaml]: the variable name should be:
sea_surface_temperature

Please change the units and the range to reflect degC

**Review by ShastriPaturi — CHANGES_REQUESTED** [review]

File: test/testinput/bufr_sfcshp.yaml (hunk 4 above)
Comment (AndrewEichmann-NOAA) [reply test/testinput/bufr_sfcshp.yaml]: Added MSST, but the results in ncdump are as follows:

>    method = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
>       0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 
>       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
>       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ;

Also, as I get a better idea how what to look for, SST0 appears in the bufr_dump but the ncdump of the result is blank. SST1 has values; is this what is needed?

File: test/testinput/bufr_sfcshp.yaml (hunk 5 above)
Comment (AndrewEichmann-NOAA) [reply test/testinput/bufr_sfcshp.yaml]: done

**Comment by ShastriPaturi** [comment]
Yes. MCSST values are 0,1,2. You’re fine since it is subsampled data.SST1 is what you should use for this ships file.You can use debufr to check for missing values.Sent from my iPhoneOn Nov 3, 2022, at 4:33 PM, AndrewEichmann-NOAA ***@***.***> wrote:﻿
@AndrewEichmann-NOAA commented on this pull request.



In test/testinput/bufr_sfcshp.yaml:
> +              hour: "*/HOUR"
+              minute: "*/MINU"
+              second: ""
+
+          stationID:
+            query: "*/RPID"
+
+          longitude:
+            query: "*/CLONH"
+
+          latitude:
+            query: "*/CLATH"
+
+          sst:
+            query: "*/SST0"
+

Added MSST, but the results in ncdump are as follows:

method = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ;

Also, as I get a better idea how what to look for, SST0 appears in the bufr_dump but the ncdump of the result is blank. SST1 has values; is this what is needed?

**Comment by AndrewEichmann-NOAA** [comment]
@ShastriPaturi Thank you for the tip - I am new at digging into BUFR files. 

Is the handling of MSST otherwise satisfactory? I've been trying to generally bring the format in line with the other yaml files in this directory, with more updates

File: test/testinput/bufr_sfcshp.yaml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        - name: "waterTemperatureMethod@MetaData"
+          source: variables/waterTemperatureMethod
+          longName: "Method of Water Temperature Measurement"
+ 
+        - name: "seaSurfaceTemperature@ObsValue"
+          coordinates: "longitude latitude"
+          source: variables/sst
+          longName: "sea_surface_temperature"
+          units: "deg C"
+          range: [-23.0, 40.0]
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshp.yaml]: Update these:
- name: sea_surface_temperature
  longName: "Sea Surface Temperature"

File: test/testinput/bufr_sfcshp.yaml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        - name: "longitude@MetaData"
+          source: variables/longitude
+          longName: "Longitude"
+          units: "degrees_east"
+          range: [-180, 180]
+
+        - name: "waterTemperatureMethod@MetaData"
+          source: variables/waterTemperatureMethod
+          longName: "Method of Water Temperature Measurement"
+
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshp.yaml]: I am not sure about this naming yet. But ok to leave it like that.

**Review by ShastriPaturi — CHANGES_REQUESTED** [review]

File: test/testinput/bufr_sfcshp.yaml (hunk 7 above)
Comment (AndrewEichmann-NOAA) [reply test/testinput/bufr_sfcshp.yaml]: Done. Is there a document of standards for these?

Comment (ShastriPaturi) [reply test/testinput/bufr_sfcshp.yaml]: @AndrewEichmann-NOAA the best place is the obs files.
Thanks for doing this.

**Comment by ShastriPaturi** [comment]
@AndrewEichmann-NOAA you also need to add a ctest for shipsu. 
Add it to this PR with a yaml name as ``bufr_sfcshipsu.yaml``

**Comment by AndrewEichmann-NOAA** [comment]
@ShastriPaturi Done. What is the difference between ships and shipsu? 
Also, there is a Location field in the netcdf output that is set to all zeros. This is present in the buoy files as well, though not specified in the yaml files. Is this a matter of concern?

**Comment by ShastriPaturi** [comment]
> @ShastriPaturi Done. What is the difference between ships and shipsu? Also, there is a Location field in the netcdf output that is set to all zeros. This is present in the buoy files as well, though not specified in the yaml files. Is this a matter of concern?

@AndrewEichmann-NOAA ships (restricted) and shipsu (unrestricted) depending on the ship the data is being measured from.
Location filed in the netcdf output should not be an issue.

File: test/testinput/bufr_sfcships.yaml [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+          source: variables/longitude
+          longName: "Longitude"
+          units: "degrees_east"
+          range: [-180, 180]
+
+        - name: "waterTemperatureMethod@MetaData"
+          source: variables/waterTemperatureMethod
+          longName: "Method of Water Temperature Measurement"
+ 
+        - name: sea_surface_temperature
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcships.yaml]: Update this as:
- name: "sea_surface_temperature@ObsValue"

File: test/testinput/bufr_sfcshipsu.yaml [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+          source: variables/longitude
+          longName: "Longitude"
+          units: "degrees_east"
+          range: [-180, 180]
+
+        - name: "waterTemperatureMethod@MetaData"
+          source: variables/waterTemperatureMethod
+          longName: "Method of Water Temperature Measurement"
+ 
+        - name: sea_surface_temperature
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshipsu.yaml]: Similar as above comment

**Review by ShastriPaturi — CHANGES_REQUESTED** [review]

File: test/testinput/bufr_sfcships.yaml [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      backend: netcdf
+      obsdataout: "./testoutput/bufr_sfcships_20180415.nc"
+
+      globals:
+        - name: "dataType@MetaData"
+          type: string
+          value: "Surface Ship"
+
+      variables:
+        - name: "datetime@MetaData"
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcships.yaml]: Rename to:
- name: "dateTime@MetaData"

File: test/testinput/bufr_sfcshipsu.yaml [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+      backend: netcdf
+      obsdataout: "./testoutput/bufr_sfcshipsu_20180415.nc"
+
+      globals:
+        - name: "dataType@MetaData"
+          type: string
+          value: "Surface Ship"
+
+      variables:
+        - name: "datetime@MetaData"
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshipsu.yaml]: Same comment as above

File: test/testinput/bufr_sfcships.yaml [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+          source: variables/timestamp
+          longName: "Datetime"
+          units: "datetime"
+
+        - name: "stationIdentification@MetaData"
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcships.yaml]: Rename to:
- name: "stationID@MetaData"

File: test/testinput/bufr_sfcshipsu.yaml [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+          source: variables/timestamp
+          longName: "Datetime"
+          units: "datetime"
+
+        - name: "stationIdentification@MetaData"
```
Comment (ShastriPaturi) [inline test/testinput/bufr_sfcshipsu.yaml]: same comment as above

**Review by ShastriPaturi — CHANGES_REQUESTED** [review]

File: test/testinput/bufr_sfcshipsu.yaml (hunk 12 above)
Comment (AndrewEichmann-NOAA) [reply test/testinput/bufr_sfcshipsu.yaml]: done

File: test/testinput/bufr_sfcships.yaml (hunk 11 above)
Comment (AndrewEichmann-NOAA) [reply test/testinput/bufr_sfcships.yaml]: done

File: test/testinput/bufr_sfcships.yaml (hunk 13 above)
Comment (AndrewEichmann-NOAA) [reply test/testinput/bufr_sfcships.yaml]: done

File: test/testinput/bufr_sfcshipsu.yaml (hunk 14 above)
Comment (AndrewEichmann-NOAA) [reply test/testinput/bufr_sfcshipsu.yaml]: done

File: test/testinput/bufr_sfcshipsu.yaml (hunk 10 above)
Comment (AndrewEichmann-NOAA) [reply test/testinput/bufr_sfcshipsu.yaml]: done

**Review by ShastriPaturi — APPROVED** [review]
Looks good @AndrewEichmann-NOAA

**Review by guillaumevernieres — APPROVED** [review]
Thanks @AndrewEichmann-NOAA . The test failure isn't related to this PR, so "all good".

**Comment by AndrewEichmann-NOAA** [comment]
@guillaumevernieres Good - was curious about that.

**Pull request merged**